src/test/operators/join_test_runner.cpp

```diff
@@ -0,0 +1,527 @@
+
+    for (auto& secondary_predicate : secondary_predicates) {
+      std::swap(secondary_predicate.column_ids.first, secondary_predicate.column_ids.second);
+      secondary_predicate.predicate_condition = flip_predicate_condition(secondary_predicate.predicate_condition);
+    }
+  }
+
+  auto to_tuple() const {
+    return std::tie(input_left, input_right, join_mode, data_type_left, data_type_right, nullable_left, nullable_right,
+                    predicate_condition, secondary_predicates);
+  }
+};
+
+bool operator<(const JoinTestConfiguration& l, const JoinTestConfiguration& r) { return l.to_tuple() < r.to_tuple(); }
+
+// Virtual interface to create a join operator
+class BaseJoinOperatorFactory {
+ public:
+  virtual ~BaseJoinOperatorFactory() = default;
+  virtual std::shared_ptr<AbstractJoinOperator> create_operator(
+      const std::shared_ptr<const AbstractOperator>& left, const std::shared_ptr<const AbstractOperator>& right,
+      const JoinMode mode, const OperatorJoinPredicate& primary_predicate,
+      const std::vector<OperatorJoinPredicate>& secondary_predicates, const JoinTestConfiguration& configuration) = 0;
+};
+
+template <typename JoinOperator>
+class JoinOperatorFactory : public BaseJoinOperatorFactory {
+ public:
+  std::shared_ptr<AbstractJoinOperator> create_operator(const std::shared_ptr<const AbstractOperator>& left,
+                                                        const std::shared_ptr<const AbstractOperator>& right,
+                                                        const JoinMode mode,
+                                                        const OperatorJoinPredicate& primary_predicate,
+                                                        const std::vector<OperatorJoinPredicate>& secondary_predicates,
+                                                        const JoinTestConfiguration& configuration) override {
+    if constexpr (std::is_same_v<JoinOperator, JoinHash>) {
+      return std::make_shared<JoinOperator>(left, right, mode, primary_predicate, secondary_predicates,
+                                            configuration.radix_bits);
+    } else {
+      return std::make_shared<JoinOperator>(left, right, mode, primary_predicate, secondary_predicates);
+    }
+  }
+};
+// Order of columns in the input tables
+const std::unordered_map<DataType, size_t> data_type_order = {
+    {DataType::Int, 0u}, {DataType::Float, 1u}, {DataType::Double, 2u}, {DataType::Long, 3u}, {DataType::String, 4u},
+};
+
+}  // namespace
+
+namespace opossum {
+
+class JoinTestRunner : public BaseTestWithParam<JoinTestConfiguration> {
+ public:
+  template <typename JoinOperator>
+  static std::vector<JoinTestConfiguration> create_configurations() {
+    auto configurations = std::vector<JoinTestConfiguration>{};
+
+    /**
+     * For each `all_*` set, the first element is the default - which is chosen to be sensible. Thus the elements are
+     * not necessarily sorted.
```

Comment:
What do "default" and "sorted mean in this context?

Reply:
Update comment to 
```
    /**
     * For each `all_*` set, the first element is used to build the default/base configuration 
     * (`JoinTestConfiguration default_configuration`). Thus the first elements are chosen to be sensible defaults.
     */
```

("sorted" meant just that: "Don't confused if it's not sorted, the elements are ordered such that the first is a good default")